src/overlays/actors/ovl_En_Dog/z_en_dog.c

```diff
@@ -22,41 +28,579 @@ const ActorInit En_Dog_InitVars =
+static void EnDog_Destroy(EnDog* this, GlobalContext* globalCtx)
+{
+    ColliderCylinderMain* collider = &this->collider; 
+    ActorCollider_FreeCylinder(globalCtx, collider);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/EnDog_Destroy.s")
+static void EnDog_FollowPath(EnDog* this, GlobalContext* globalCtx)
+{
+    s16ArrEntry behaviors[] = { 3, 5, 2 };
+    s16ArrEntry unused[] = { 40, 80, 20 };
+    f32 speed;
+    s32 frame;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB6C4.s")
+    if (EnDog_CanFollow(this, globalCtx) == 1)
+    {
+        this->actionFunc = EnDog_FollowLink;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB858.s")
+    if(DECR(this->behaviorTimer))
+    {
+        if (this->nextBehavior == DOG_WALK)
+        {
+            speed = 1.0f;
+        }
+        else
+        {
+            speed = 4.0f;
+        }
+        Math_SmoothScaleMaxMinF(&this->actor.speedXZ, speed, 0.4f, 1.0f, 0.0f);
+        EnDog_Orient(this, globalCtx);
+        this->actor.shape.rot = this->actor.posRot.rot;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB940.s")
+        // Used to change between two text boxes for Richard's owner in the Market Day scene
+        // depending on where he is on his path. En_Hy checks these event flags.
+        if (this->waypoint < 9)
+        {
+            // Richard is close to her, text says something about his coat
+            gSaveContext.event_inf[3] |= 1;
+        }
+        else
+        {
+            // Richard is far, text says something about running fast
+            gSaveContext.event_inf[3] &= ~1;
+        }
+    }
+    else
+    {
+        frame = globalCtx->state.frames % 3;
+        this->nextBehavior = behaviors[frame].entry;
+        // no clue why they're using the action id to calculate timer. possibly meant to use the unused array?
+        this->behaviorTimer = Math_Rand_S16Offset(60, behaviors[frame].unk_2);
+        this->actionFunc = EnDog_ChooseMovement;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FBABC.s")
+static void EnDog_ChooseMovement(EnDog* this, GlobalContext* globalCtx)
+{
+    if (EnDog_CanFollow(this, globalCtx) == 1)
+    {
+        this->actionFunc = EnDog_FollowLink;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FBB64.s")
+    if(DECR(this->behaviorTimer) == 0)
+    {
+        this->behaviorTimer = Math_Rand_S16Offset(200, 100);
+        if (globalCtx->state.frames % 2)
+        {
+            this->nextBehavior = DOG_WALK;
+        }
+        else
+        {
+            this->nextBehavior = DOG_RUN;
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FBC6C.s")
+        if (this->nextBehavior == DOG_RUN)
+        {
+            this->behaviorTimer = (this->behaviorTimer / 2);
```

Comment:
I think I would do this if it matched.
```suggestion
            this->behaviorTimer /= 2;
```